### PR TITLE
Kopia repository server controller deployment

### PR DESCRIPTION
**Pull request by kale-amruta**
Kopia repository server deployment

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #issue-number

## Test Plan

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: helm/kanister-operator/values.yaml
Comment (ankitjain235) [inline helm/kanister-operator/values.yaml:9]:
```suggestion
  registry: ghcr.io/kanisterio
  name: repo-server-controller
```
I think we should break down registry and name into two separate fields. This would allow using a different registry if required.

File: helm/kanister-operator/templates/deployment.yaml
Comment (ankitjain235) [inline helm/kanister-operator/templates/deployment.yaml:43]:
```suggestion
        image: {{ .Values.repositoryServerImage.registry }}/{{ .Values.repositoryServerImage.name }}:{{ .Values.repositoryServerImage.tag }}
```
If we split image into registry and name, this would be required.

File: helm/kanister-operator/templates/_helpers.tpl
Comment (ankitjain235) [inline helm/kanister-operator/templates/_helpers.tpl:31]: I cannot see this field `repoServerControllerName` being set in values.yaml. Do we need to add field with empty value here?

File: docker/repositoryserver-controller/Dockerfile
Comment (ankitjain235) [inline docker/repositoryserver-controller/Dockerfile:1]: I see this was updated recently. Let's make sure we update to latest before merging.

**Review by ankitjain235 — COMMENTED** [review]
Mostly looks good, added some minor suggestions.

File: helm/kanister-operator/values.yaml
Comment (kale-amruta) [reply helm/kanister-operator/values.yaml:9]: Wouldnt that be inconsistent with current kanister controller image format?

File: helm/kanister-operator/templates/_helpers.tpl
Comment (kale-amruta) [reply helm/kanister-operator/templates/_helpers.tpl]: Ah, that was missing, thank you

File: .goreleaser.yml
Comment (viveksinghggits) [inline .goreleaser.yml:57]: Should we name the directory under docker, `repo-server-controller`  as well? We might get confused what has repository and what has repo, so is it better to just have `repo-` everywhere?

File: helm/kanister-operator/values.yaml
Comment (viveksinghggits) [inline helm/kanister-operator/values.yaml:8]: should this be named `repositoryServerControllerImage`?
and line 32 as well.

**Review by viveksinghggits — COMMENTED** [review]
LGTM. I will approve once the questions are answered.

File: Makefile.kubebuilder
Comment (viveksinghggits) [inline Makefile.kubebuilder:70]: Should we remove this extra new line?

File: build/generate_crds.sh
Comment (viveksinghggits) [inline build/generate_crds.sh:34]: Should we remove this extra newline?

**Review by viveksinghggits — APPROVED** [review]

File: Makefile.kubebuilder
Comment (ankitjain235) [inline Makefile.kubebuilder:4]: Same here as well, do we need both copyrights?

File: helm/kanister-operator/values.yaml
Comment (ankitjain235) [inline helm/kanister-operator/values.yaml:11]:
```suggestion
  tag: 0.89.0
```
Or whichever is latest at the time of merging.

File: build/generate_crds.sh
Comment (ankitjain235) [inline build/generate_crds.sh:3]: Should we add copyright here since we already have the copyright below?

**Review by ankitjain235 — COMMENTED** [review]
LGTM overall.

File: helm/kanister-operator/values.yaml
Comment (kale-amruta) [reply helm/kanister-operator/values.yaml:11]: will update this with latest when we will merge this PR

File: helm/kanister-operator/values.yaml
Comment (viveksinghggits) [inline helm/kanister-operator/values.yaml:41]:
```suggestion
    name: ''
```

Would this work?

File: build/generate_crds.sh
Comment (viveksinghggits) [inline build/generate_crds.sh:1]: does this generate all the Kanister CRDs or just the RepositoryServer CRD?

File: Makefile.kubebuilder
Comment (viveksinghggits) [inline Makefile.kubebuilder:1]: are we using targets from this Makefile anywhere?
The only useful targets that I see, are `repo-server-build`, `repo-server-install-crd` and `repo-server-uninstall-crd`.

 ### repo-server-build
If we are using goreleaser to build the binary and since goreleaser is being used to build other binaries I think it's not a good idea to do that using this other makefile as well.

### repo-server-install-crd, repo-server-uninstall-crd
When, by who is this target going to be used?

Comment (kale-amruta) [reply Makefile.kubebuilder:1]: These targets are for development purposes/ anybody who wants to create/delete crd

Comment (viveksinghggits) [reply Makefile.kubebuilder:1]: if we eventually plan to distribute the CRD like we are doing for the CRDs I don't think these targets are required.

File: build/generate_crds.sh
Comment (kale-amruta) [reply build/generate_crds.sh:1]: just the repository server CRD

Comment (kale-amruta) [reply build/generate_crds.sh:1]: just the repository server CRD

File: helm/kanister-operator/templates/deployment.yaml
Comment (PrasadG193) [inline helm/kanister-operator/templates/deployment.yaml:45]: Let's make sure we don't deploy the repo server controller by default as of now. Let's keep this behind a feature flag till we have it ready for GA

File: helm/kanister-operator/values.yaml
Comment (ankitjain235) [inline helm/kanister-operator/values.yaml:9]: A note to change this before merging.

File: docker/repo-server-controller/Dockerfile
Comment (ankitjain235) [inline docker/repo-server-controller/Dockerfile:5]: Why do we need to run this command?

**Review by ankitjain235 — COMMENTED** [review]
LGTM besides minor comment.

File: helm/kanister-operator/values.yaml
Comment (PrasadG193) [inline helm/kanister-operator/values.yaml:43]: Let's set the default values instead of keeping them blank. So that the default values will be rendered to generated helm chart

File: docker/repo-server-controller/Dockerfile
Comment (kale-amruta) [reply docker/repo-server-controller/Dockerfile:5]: Even I am unsure of that, I followed the dockerfile of kanister controller

Comment (kale-amruta) [reply docker/repo-server-controller/Dockerfile:5]: This PR has details about why it was updated in kanister controller, since we are using same base image for repo server controller, I kept it as it is:
https://github.com/kanisterio/kanister/pull/949

File: build/generate_crds.sh
Comment (PrasadG193) [inline build/generate_crds.sh:31]: - How are we generating Kanister CRDs currently?
- What changes would it need to generate Kanister CRDs with controller-gen?

I believe we should follow a common pattern for all the CRDs?

Comment (kale-amruta) [reply build/generate_crds.sh:31]: @PrasadG193 This generates all the CRDs under package `github.com/kanisterio/kanister/pkg/apis/cr/v1alpha1`. I dont see any target in the makfile which generates kanister CRDs apart from the one I have written, We have `make codegen`, but that  just generates the code and not the CRDs. Hence I think this would be the right target to generate all the CRDs

Comment (kale-amruta) [reply build/generate_crds.sh:31]: Let me move the target to the main Makefile rather than kubebuilder`s makefile

Comment (kale-amruta) [reply build/generate_crds.sh:31]: Done. let me know if we should merge this PR now?

File: Makefile
Comment (PrasadG193) [inline Makefile:78]: Should we use the latest one - https://github.com/kubernetes-sigs/controller-tools/releases

**Review by PrasadG193 — COMMENTED** [review]
LGTM. Just have minor comment around if we should switch to latest version of controller-tools.

**Review by PrasadG193 — APPROVED** [review]

File: Makefile
Comment (kale-amruta) [reply Makefile]: Done

**Pull request merged**
